### PR TITLE
Document Symbol Provider and Definition Provider

### PR DESCRIPTION
**Pull request by TheColorRed**
Added 2 providers:

**Document Symbol Provider**
Shows a list of Classes/Methods/Properties/etc. (Symbols) in the current document.
- Triggered by <kbd>Ctrl+Shift+O</kbd>

**Definition Provider**
Supports "Go to definition"
- Triggered by <kbd>F12</kbd> and <kbd>Ctrl+Click</kbd>

### CODE REVIEW

File: server/src/server.ts [hunk 1]
```diff
@@ -141,6 +141,229 @@ connection.onRequest(buildObjectTreeForDocument, (requestObj) =>
...
+                symbol.kind = SymbolKind.Class;
+                symbol.startLine = classNode.startPos.line;
+                symbol.startChar = classNode.startPos.col;
+                symbol.endLine = classNode.endPos.line;
+                symbol.endChar = classNode.endPos.col;
+                symbol.path = item.path;
+                symbols.push(symbol);
+            }
+            // Search the methods within the classes
+            classNode.methods.forEach(methodNode => {
```
Comment (nevadascout) [inline server/src/server.ts]: What happens if two classes have a method with the same name?

File: server/src/hvy/treeBuilder.ts [hunk 2]
```diff
@@ -222,8 +228,13 @@ export class TreeBuilder
                             });
 
                             var symbolCache = new FileSymbolCache();
-                            symbolCache.name = methodNode.name;
+                            symbolCache.name      = methodNode.name;
```
Comment (nevadascout) [inline server/src/hvy/treeBuilder.ts]: Can you fix the whitespace in these changes please?

**Comment by nevadascout** [comment]
Does go to definition work for symbols outside the current document?

File: server/src/server.ts (hunk 1 above)
Comment (TheColorRed) [inline server/src/server.ts]: It pops open a window with a list of all the definitions

File: server/src/hvy/treeBuilder.ts (hunk 2 above)
Comment (TheColorRed) [inline server/src/hvy/treeBuilder.ts]: White space fixed

**Comment by TheColorRed** [comment]
Yeah, I was able to go to a stub symbol. So basically as long as it is defined in the workspace tree it is accessible.

**Comment by phptek** [comment]
Hi and thanks for all the hard-work on this! Any idea when this may make it into a stable enough release such that I can install the module with this feature, from within VSCode?

**Comment by nevadascout** [comment]
Hey @phptek thanks for your kind comment!

@ichiriac and I are working on an overhaul of Crane using a new parser which should allow us to do this and many other things that I originally planed for Crane including go-to-definition, file symbols, workspace symbols and much more.

We're aiming to have the first version of this overhaul done within a few weeks, but at the moment it's too early to give any exact dates for when we expect it to be completed.

You can see our progress on the [feature/reflection-rewrite](https://github.com/HvyIndustries/crane/tree/feature/reflection-rewrite) branch

**Pull request merged**